crates/polars-stream/src/nodes/io_sources/parquet/init.rs

```diff
@@ -135,100 +130,48 @@ impl ParquetSourceNode {
 
             row_group_data_fetcher.slice_range = slice_range;
 
-            // Ensure proper backpressure by only polling the buffered iterator when a wait group
-            // is free.
-            let mut wait_groups = (0..num_pipelines)
-                .map(|index| IndexedWaitGroup::new(index).wait())
-                .collect::<FuturesUnordered<_>>();
-
-            let mut df_stream = row_group_data_fetcher
-                .into_stream()
-                .map(|x| async {
-                    match x {
-                        Ok(handle) => handle.await.unwrap(),
-                        Err(e) => Err(e),
-                    }
-                })
-                .buffered(row_group_prefetch_size)
-                .map(|x| async {
-                    let row_group_decoder = row_group_decoder.clone();
-
-                    match x {
-                        Ok(row_group_data) => {
-                            async_executor::spawn(TaskPriority::Low, async move {
-                                row_group_decoder.row_group_data_to_df(row_group_data).await
-                            })
-                            .await
-                        },
-                        Err(e) => Err(e),
-                    }
-                })
-                .buffered(
-                    // Because we are using an ordered buffer, we may suffer from head-of-line blocking,
-                    // so we add a small amount of buffer.
-                    num_pipelines + 4,
-                );
-
             let morsel_seq_ref = &mut MorselSeq::default();
-            let mut dfs = VecDeque::with_capacity(1);
+            let mut prefetches = VecDeque::with_capacity(row_group_prefetch_size);
+            let mut decoded_dfs = VecDeque::with_capacity(1);
 
-            'main: loop {
-                let Some(mut indexed_wait_group) = wait_groups.next().await else {
-                    break;
-                };
-
-                while dfs.is_empty() {
-                    let Some(v) = df_stream.next().await else {
-                        break 'main;
+            loop {
+                while prefetches.len() < row_group_prefetch_size {
```

Comment:
I don't think this is ideal - we are now forcibly filling `prefetches` to `row_group_prefetch_size` at the start of every loop - there could be existing `prefetches` at the start that have morsels ready to send but get blocked by this.

You can see `.buffered()` from futures-util breaks when it gets a `Poll::Pending` from the stream -

https://github.com/rust-lang/futures-rs/blob/1e052816b09890925cfdfcbe8d390cdaae5e4c38/futures-util/src/stream/stream/buffered.rs#L68

Reply:
> I don't think this is ideal - we are now forcibly filling prefetches to row_group_prefetch_size at the start of every loop - there could be existing prefetches at the start that have morsels ready to send but get blocked by this.

How can they get blocked? We only start the prefetches here, not fully await them. Besides, that's the whole point of a prefetch - that you start prefetching things even though you still have data available to you right now.

Reply:
When we call `row_group_data_fetcher.next().await`, internally it will call `init_next_file_state().await` if we have depleted the row groups of the current file, so we end up blocking until the metadata of the next file is ready -

https://github.com/pola-rs/polars/blob/d68689f6eb6b613e8a354248c1e2fe79080c29b3/crates/polars-stream/src/nodes/io_sources/parquet/row_group_data_fetch.rs#L63